### DISCUSSION TITLE
Pass arguments to an app factory

### DISCUSSION THREAD

**Discussion by adamlwgriffiths** [Kludex/uvicorn#1644] [Q&A]
I'm wanting to pass the path to a configuration file to my server via the cli.
My code looks something like this:
```
def create_app(config_filename):
    # make the app
    # configure myself
    return app

if __name__ == '__main__':
    uvicorn.run('my_app:create_app', factory=True, ...)
```
Ie. the app the builds and configures itself.

I'm simply wanting to pass arguments to the factory method that uvicorn calls.
I don't want uvicorn to attempt to load or parse these arguments.

I've looked around the documentation and in regards to configuration it's very unclear on how to do this.
There is mention of a `--env-file` via cli, and the Config class supports an `env_file` argument.
But both of these are parsed and somehow passed in.
I've already got a configuration file, I _really_ don't want to write another one just as a way tell it to load the original.

How can I go about this?

**Comment by Kludex** [comment]
You can run `uvicorn` programmatically: https://www.uvicorn.org/#config-and-server-instances.

**Reply by SaravananSathyanandhaQC** [reply]
@Kludex how would this work (a) with multiple workers where you need to use a factory, and (b) when running with gunicorn? I proposed #2445 to support this, but maybe I misunderstood this comment and there is a way to do this with multiple workers

**Comment by adamlwgriffiths** [comment]
I've seen that page and it didn't appear to help as it doesn't demonstrate how to pass in arguments to an application factory function.
It instead implies that the function there must adhere to the `(scope, receive, send)` signature.

**Comment by JimiC** [comment]
The answer is you can't. Simply because it's not supported.
https://github.com/encode/uvicorn/blob/3d26ab46597b368ba45d68e8b0f65342ede8d2e6/uvicorn/config.py#L440

You need to use a workaround. In my case I used manual dependencies creation.

Another solution could be using dependency injection via a DI package (like [dependency-injector](https://python-dependency-injector.ets-labs.org/)). Haven't tried it yet.

**Comment by hchargois-ipsos** [comment]
PR #2445 implements the required feature for this to work.
